torch/_inductor/freezing.py

```diff
@@ -0,0 +1,208 @@
+import itertools
+import weakref
+from typing import List, Optional, Tuple
+
+import torch
+import torch.utils._pytree as pytree
+from . import config
+
+
+def replace_node_with_constant(gm, node, constant):
+    g = gm.graph
+
+    i = 0
+    while True:
+        qualname = f"_frozen_param{i}"
+        if not hasattr(gm, qualname):
+            break
+        i += 1
+
+    with g.inserting_before(node):
+        new_input_node = g.create_node("get_attr", qualname, (), {})
+        node.replace_all_uses_with(new_input_node)
+        new_input_node.meta.update(node.meta)
+        g.erase_node(node)
+
+    # needed to suppress `does not reference an nn.Module, nn.Parameter, or buffer` warning
+    gm.register_buffer(qualname, constant)
+    setattr(gm, qualname, constant)
+
+
+def replace_params_with_constants(gm, flat_params, fw_metadata) -> List[int]:
+    """
+    Replaces the parameters of a PyTorch GraphModule with constants wherever possible.
+
+    Returns a list of indices representing the input parameters that were not converted to constants.
+    """
+
+    params = [node for node in gm.graph.nodes if node.op == "placeholder"]
+    fake_inp_nodes = [node for (_, node) in zip(flat_params, params)]
+
+    g = gm.graph
+
+    preserved_arg_indices = []
+    aliased_input_args = [
+        out_info.base_idx
+        for out_info in fw_metadata.output_info
+        if out_info.base_idx is not None
+    ]
+
+    for i, (real_input, node) in enumerate(zip(flat_params, fake_inp_nodes)):
+        if i in fw_metadata.mutated_inp_indices or aliased_input_args:
+            preserved_arg_indices.append(i)
+            continue
+
+        replace_node_with_constant(gm, node, real_input)
+
+    # add on non param inputs
+    preserved_arg_indices.extend(range(len(flat_params), len(params)))
+
+    # is this necessary ?
+    gm.recompile()
+    return preserved_arg_indices
+
+
+@torch.utils._python_dispatch._disable_current_modes()
+def constant_fold(gm):
+    unknown_value = object()
+
+    node_replacements = {}
+
+    class ConstantFolder(torch.fx.Interpreter):
+        def run_node(self, node):
+            args, kwargs = self.fetch_args_kwargs_from_env(node)
+            if unknown_value in pytree.tree_flatten((args, kwargs))[0]:
+                return unknown_value
+
+            # All mutations should either be removed or on inputs which we did not make constant
+            if (
+                isinstance(node.target, torch._ops.OpOverload)
+                and torch.Tag.nondeterministic_seeded in node.target.tags
+            ):
+                return unknown_value
+
+            out = super().run_node(node)
+
+            # TODO - remove constant from node_replacement when it has no uses
+            if node.op != "get_attr" and isinstance(out, torch.Tensor):
+                node_replacements[node] = out
```

Comment:
@eellison Do you think is it reasonable adding some check here to filter nodes we don't want to do the constant folding? For example, take quantization conv as example, there is a pattern `fp32_weight->quant_per_channel->dequant_per_channel->conv`. Although `fp32_weight->quant_per_channel->dequant_per_channel` can do constant folding into new fp32 weight, however, we want to keep `dequant_per_channel` node in the graph. Then `fp32_weight->quant_per_channel` can be constant folding into a int8_weight, and `dequant_per_channel->conv` can be further lowering into a `int8_conv` node. Also cc @jgong5 @Xia-Weiwen.

Reply:
Yea, makes sense, we can deal with this and the other quantization issue in a follow up